src/test/java/com/exactpro/th2/taskutils/FutureTrackerTest.java

```diff
@@ -63,104 +104,334 @@ private CompletableFuture<Integer> chainFuture (CompletableFuture<Integer> futur
+            assertLessDuration(NO_DELAY_MILLIS, () -> assertThat(futureTracker.track(extraFuture, DELAY_DURATION)).isTrue());
 
-        long start = System.nanoTime();
-        futureTracker.track(getFutureWithException());
-        futureTracker.awaitRemaining();
+            assertThat(futureTracker.remaining()).isEqualTo(0);
+        }
+
+        @ParameterizedTest
+        @ValueSource(ints = {0, LIMIT})
+        public void testTrackFutureAfterStop(int limit) throws InterruptedException {
+            FutureTracker<Integer> futureTracker = FutureTracker.create(limit);
+
+            assertThat(futureTracker.isEnabled()).isTrue();
+            assertLessDuration(NO_DELAY_MILLIS, futureTracker::awaitRemaining);
+            assertThat(futureTracker.isEnabled()).isFalse();
 
-        long actualTrackingMillis = (System.nanoTime() - start)/1_000_000;
-        Assertions.assertThat(actualTrackingMillis).isLessThan(expectedTrackingMillis);
+            StartableRunnable waitingRunnable = StartableRunnable.of(new SleepingRunnable(DELAY_MILLIS));
+            CompletableFuture<Integer> curFuture = getFutureWithDelay(waitingRunnable);
+
+            assertLessDuration(NO_DELAY_MILLIS, () -> {
+                IllegalStateException exception = assertThrows(IllegalStateException.class, () -> futureTracker.track(curFuture));
+                assertThat(exception.getMessage()).isEqualTo("Future tracker is already disabled");
+            });
+            assertLessDuration(NO_DELAY_MILLIS, () -> {
+                IllegalStateException exception = assertThrows(IllegalStateException.class, () -> futureTracker.track(curFuture, DELAY_DURATION));
+                assertThat(exception.getMessage()).isEqualTo("Future tracker is already disabled");
+            });
+        }
     }
 
-    @Test
-    public void testTracking5Futures() {
-        long expectedTrackingTime = 5 * DELAY_MILLIS;
+    @Nested
+    class Unlimited {
+        private final FutureTracker<Integer> futureTracker = FutureTracker.createUnlimited();
 
-        StartableRunnable waitingRunnable = StartableRunnable.of(new SleepingRunnable(DELAY_MILLIS));
+        @Test
+        public void testLimit() {
+            assertThat(futureTracker.limit()).isLessThanOrEqualTo(0);
+            assertThat(futureTracker.hasLimit()).isFalse();
+        }
+    }
 
-        FutureTracker<Integer> futureTracker = new FutureTracker<>();
+    @Nested
+    class Limited {
+        private final FutureTracker<Integer> futureTracker = FutureTracker.create(LIMIT);
 
-        CompletableFuture<Integer> lastFuture = null;
-        for (int i = 0; i < 5; i ++) {
-            CompletableFuture<Integer> curFuture = getFutureWithDelay(waitingRunnable);
-            if (i != 0) {
-                curFuture = chainFuture(lastFuture);
+        @Test
+        public void testLimit() {
+            assertThat(futureTracker.limit()).isGreaterThan(0);
```

Comment:
question: Why don't we check that the limit is equal to `LIMIT`? We know the exact limit in advance

Reply:
We check that future tracker has positive limit for this mode. LMIT value should not affect this check.

Reply:
Hmm... do you say if I create a future tracker with limit 5, is it okay if the `limit` call returns 3 instead of 5? Right now it sounds like it is okay

Reply:
Yes, I would like to check mode - positive limit, instead of limit value

Reply:
Okay, up to you. But it still does not make sense to me